### PR TITLE
vpi defname

### PR DESCRIPTION
**Pull request by AndrewNolte**
(loosely) depends on https://github.com/verilator/verilator/issues/3836

issue: https://github.com/verilator/verilator/issues/3906

### CODE REVIEW

File: include/verilated.h [hunk 1]
```diff
@@ -608,17 +608,22 @@ class VerilatedScope final {
     int8_t m_timeunit = 0;  // Timeunit in negative power-of-10
     Type m_type = SCOPE_OTHER;  // Type of the scope
 
+    const char* m_defNamep = nullptr;  // defName (module name, not instance name)
```
Comment (wsnyder) [inline include/verilated.h:611]: Please remove newline above, thanks.

File: include/verilated.h [hunk 2]
```diff
@@ -608,17 +608,22 @@ class VerilatedScope final {
...
-                   const char* identifier, int8_t timeunit, const Type& type) VL_MT_UNSAFE;
+                   const char* identifier, int8_t timeunit, const Type& type,
+                   const char* defName) VL_MT_UNSAFE;
     void exportInsert(int finalize, const char* namep, void* cb) VL_MT_UNSAFE;
     void varInsert(int finalize, const char* namep, void* datap, bool isParam,
                    VerilatedVarType vltype, int vlflags, int dims, ...) VL_MT_UNSAFE;
     // ACCESSORS
     const char* name() const VL_MT_SAFE_POSTINIT { return m_namep; }
     const char* identifier() const VL_MT_SAFE_POSTINIT { return m_identifierp; }
+
```
Comment (wsnyder) [inline include/verilated.h:625]:
```suggestion
```

File: include/verilated_vpi.cpp [hunk 3]
```diff
@@ -277,6 +277,8 @@ class VerilatedVpioScope VL_NOT_FINAL : public VerilatedVpio {
     const VerilatedScope* scopep() const { return m_scopep; }
     const char* name() const override { return m_scopep->name(); }
     const char* fullname() const override { return m_scopep->name(); }
+
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:280]:
```suggestion
```

File: test_regress/t/t_vpi_module.cpp [hunk 4]
```diff
@@ -192,6 +192,7 @@ int main(int argc, char** argv) {
     const std::unique_ptr<VM_PREFIX> topp{new VM_PREFIX{contextp.get(),
                                                         // Note null name - we're flattening it out
                                                         ""}};
+    contextp->scopesDump();
```
Comment (wsnyder) [inline test_regress/t/t_vpi_module.cpp:195]: Remove, it's already below, you can use --verbose when running a test to see it.

File: test_regress/t/t_vpi_var.cpp [hunk 5]
```diff
@@ -593,6 +595,8 @@ int _mon_check_putget_str(p_cb_data cb_data) {
         };
     } else {
         // setup and install
+        const char* p;
```
Comment (wsnyder) [inline test_regress/t/t_vpi_var.cpp:598]:
```suggestion
        const char* p = nullptr;
```

Just for safety.

File: src/V3EmitCSyms.cpp [hunk 6]
```diff
@@ -851,7 +876,10 @@ void EmitCSyms::emitSymImp() {
             putsQuoted(protect(scopeDecodeIdentifier(it->second.m_prettyName)));
             puts(", ");
             puts(cvtToStr(it->second.m_timeunit));
-            puts(", VerilatedScope::" + it->second.m_type + ");\n");
+            puts(", VerilatedScope::" + it->second.m_type);
+            puts(", ");
+            putsQuoted(it->second.m_defName);
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:881]:
```suggestion
            putsQuoted(protect(it->second.m_defName));
```

File: src/V3EmitCSyms.cpp [hunk 7]
```diff
@@ -296,10 +319,11 @@ class EmitCSyms final : EmitCBaseVisitor {
...
                 = (nodep->origModName() == "__BEGIN__") ? "SCOPE_OTHER" : "SCOPE_MODULE";
             const string name = nodep->scopep()->shortName() + "__DOT__" + nodep->name();
-            const string name_dedot = AstNode::dedotName(name);
+            const string name_pretty = AstNode::vpiName(name);
             const int timeunit = m_modp->timeunit().powerOfTen();
-            m_vpiScopeCandidates.insert(
-                std::make_pair(name, ScopeData(scopeSymString(name), name_dedot, timeunit, type)));
+            m_vpiScopeCandidates.insert(std::make_pair(
+                name, ScopeData(scopeSymString(name), name_pretty, timeunit, type,
+                                type == "SCOPE_MODULE" ? nodep->origModName() : "")));
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:326]:
```suggestion
                name, ScopeData{scopeSymString(name), name_pretty, timeunit, type,
                                type == "SCOPE_MODULE" ? nodep->origModName() : ""}));
```

Might need some other minor changes when upgrade to {}.

File: src/V3EmitCSyms.cpp [hunk 8]
```diff
@@ -44,12 +44,15 @@ class EmitCSyms final : EmitCBaseVisitor {
         const string m_prettyName;
         const int m_timeunit;
         string m_type;
+
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:47]:
```suggestion
```

File: src/V3EmitCSyms.cpp [hunk 9]
```diff
@@ -156,6 +159,25 @@ class EmitCSyms final : EmitCBaseVisitor {
         string out = scpname;
         // Remove hierarchy
         string::size_type pos = out.rfind('.');
+
+        // If there's more than one ident and an escape, find the true last ident
+        if (pos != string::npos && scpname.find('\\') != string::npos) {
+            int i = 0;
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:165]:
```suggestion
            size_t i = 0;
```

File: src/V3EmitCSyms.cpp [hunk 10]
```diff
@@ -156,6 +159,25 @@ class EmitCSyms final : EmitCBaseVisitor {
...
+            // always makes progress
+            while (i < scpname.length()) {
+                if (scpname[i] == '\\') {
+                    while (i < scpname.length() && scpname[i] != ' ') { i++; }
+                    i++;  // proc ' ', it should always be there. Then grab '.' on next cycle
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:170]:
```suggestion
                    ++i;  // proc ' ', it should always be there. Then grab '.' on next cycle
```

File: src/V3EmitCSyms.cpp [hunk 11]
```diff
@@ -156,6 +159,25 @@ class EmitCSyms final : EmitCBaseVisitor {
...
+                } else {
+                    while (i < scpname.length() && scpname[i] != '.') { i++; }
+                    if (i < scpname.length()) {
+                        pos = i;
+                        i++;
```
Comment (wsnyder) [inline src/V3EmitCSyms.cpp:175]:
```suggestion
                        pos = i++;
```

**Comment by AndrewNolte** [comment]
I'm noticing that some modules get the type "SCOPE_OTHER", instead of scope module, is this a bug? Or is it sometimes supposed to do this? It's coming from the first emplace in visit(AstScopeName). Is this one of those data objs that's supposed to get overwritten in varHierarchyScopes?

**Comment by wsnyder** [comment]
> I'm noticing that some modules get the type "SCOPE_OTHER", instead of scope module, is this a bug?
I think this is a bug.  Perhaps these were inlined?  They should still look like Modules to the VPI.

**Comment by AndrewNolte** [comment]
I care more about the defname than the module/other type though, so maybe I could have it copy over all of the defnames on matching names, and leave the type copyover where it was. I'm not sure if the module is even getting pulled in though. I think it has to do with how deep the public wires are.

**Comment by AndrewNolte** [comment]
I was able to get the module name with m_modp->origName() in the astscopename visitor, but some of these are wrong. From what I can tell though every symbol emmited from this line is a module

**Pull request closed without merging**